src/components/colorbar/connect.js

```diff
@@ -0,0 +1,85 @@
+/**
+* Copyright 2012-2018, Plotly, Inc.
+* All rights reserved.
+*
+* This source code is licensed under the MIT license found in the
+* LICENSE file in the root directory of this source tree.
+*/
+
+
+'use strict';
+
+var isNumeric = require('fast-isnumeric');
+
+var aggNums = require('../../lib').aggNums;
+var Colorscale = require('../colorscale');
+var drawColorbar = require('./draw');
+
+/**
+ * connectColorbar: create a colorbar from a trace, using its module to
+ *   describe the connection.
+ *
+ * @param {DOM element} gd
+ *
+ * @param {Array} cd
+ *   calcdata entry for this trace. cd[0].trace is the trace itself, and the
+ *   colorbar object will be stashed in cd[0].t.cb
+ *
+ * @param {object|function} moduleOpts
+ *   may be a function(gd, cd) to override the standard handling below. If
+ *   an object, should have these keys:
+ * @param {Optional(string)} moduleOpts.container
+ *   name of the container inside the trace where the colorbar and colorscale
+ *   attributes live (ie 'marker', 'line') - omit if they're at the trace root.
+ * @param {string} moduleOpts.min
+ *   name of the attribute holding the value of the minimum color
+ * @param {string} moduleOpts.max
+ *   name of the attribute holding the value of the maximum color
+ * @param {Optional(string)} moduleOpts.vals
+ *   name of the attribute holding the (numeric) color data
+ *   used only if min/max fail. May be omitted if these are always
+ *   pre-calculated.
+ */
+module.exports = function connectColorbar(gd, cd, moduleOpts) {
+    if(typeof moduleOpts === 'function') return moduleOpts(gd, cd);
+
+    var trace = cd[0].trace;
+    var cbId = 'cb' + trace.uid;
+    var containerName = moduleOpts.container;
+    var container = containerName ? trace[containerName] : trace;
+
+    gd._fullLayout._infolayer.selectAll('.' + cbId).remove();
+    if(!container || !container.showscale) return;
+
+    var zmin = container[moduleOpts.min];
+    var zmax = container[moduleOpts.max];
+
+    var valAttr = moduleOpts.vals;
+    var vals;
+    if(Array.isArray(valAttr)) {
+        for(var i = 0; i < valAttr.length; i++) {
+            vals = container[valAttr[i]];
+            if(vals) break;
+        }
+    }
+    else vals = container[valAttr];
+
+    if(vals) {
```

Comment:
(Non-blocking) I'm curious to see what would happen (i.e. which tests would fail if any) if we drop the logic around vals here. cmin/cmax or zmin/zmax should always be computed during the calc step which should happen before this block.

Reply:
yeah, looks like calculating the range while drawing the colorbar is obsolete, good call! Though I can't seem to run the image tests all the way through locally anymore, so we'll see in a few minutes -> https://github.com/plotly/plotly.js/pull/2681/commits/f9fbbc647cad99497514c8a575c111b7222f8d73

Reply:
![image](https://user-images.githubusercontent.com/6675409/40785298-2c983cc2-64b6-11e8-85b2-2691cccdf042.png)

:white_check_mark: nice!